### ISSUE TITLE
GSoC Idea: camera calibration wizard

### ISSUE THREAD

**Issue by s-trinh** [opencv/opencv#21132]
Camera calibration is probably the most common topic taught and needed in computer vision.
Yet I don't think most of the people know how to properly calibrate a camera:
- blurry images
- same viewpoints
- calibration board printed on a sheet of paper
- bad illumination
- ...

There are also some misconceptions I believe. For instance:
- I don't think taking fronto parallel calibration views is a good think with the internal calibration method, see:
![image](https://user-images.githubusercontent.com/8035162/143599064-dc8221ae-6efa-4ab5-aae6-5a7f4cc0b411.png)
- people usually takes lots of images while I don't think it is really needed, with the idea that the more the images we have, the better is the quality of the parameters estimation

This page has some camera calibration advices: https://www.dlr.de/rm/en/desktopdefault.aspx/tabid-3925/6084_read-9196/

An interactive tool to assist the user about how to correctly position the calibration target in front of the camera could be an idea for a potential GSoC student?
Some potential features for the calibration tool:
- support different calibration patterns?
- mono + stereo calibration?
- assist the user for the next board viewpoint to use?, show where to position the board in the image
- don't use blurry images?

Some references:
- [AprilCal: Assisted and repeatable camera calibration](https://april.eecs.umich.edu/papers/details.php?name=richardson2013iros):
  - [presentation](https://april.eecs.umich.edu/media/aprilcal_slides_iros2013.pdf)
  - [video](https://april.eecs.umich.edu/media/aprilcal/aprilcal_1080p.mov)
  - [Wiki](https://april.eecs.umich.edu/wiki/Camera_suite)
  - [Java code?](https://april.eecs.umich.edu/software/java.html)
- [Calibration Wizard: A Guidance System for Camera Calibration Based on Modelling Geometric and Corner Uncertainty](https://hal.inria.fr/hal-02368999/document):
  - [poster](https://pengsongyou.github.io/files/iccv19_poster.pdf)
  - [paper](https://hal.inria.fr/hal-02368999/document)
  - [Github code](https://github.com/pengsongyou/CalibrationWizard)
- [Efficient Pose Selection for Interactive Camera Calibration](https://arxiv.org/pdf/1907.04096.pdf):
  - [Github code](https://github.com/paroj/pose_calib)

**Comment by alalek** [comment]
Please take a look on this builtin app: https://github.com/opencv/opencv/tree/4.x/apps/interactive-calibration

**Comment by s-trinh** [comment]
Something like this video (https://april.eecs.umich.edu/media/aprilcal/aprilcal_1080p.mov) would be ideal (assisted calibration for reproducible results).

Anybody is welcome to comment on this idea.

**Comment by s-trinh** [comment]
**/cc** @vpisarev since you have written the [OE 20. Module Calibration](https://github.com/opencv/opencv/wiki/OE-20.-Module-Calibration)

Some various ideas below.

---

## Refresh of the existing calibration tutorials / tools?

- add a default `ChArUco` calibration board image similar to the https://github.com/opencv/opencv/blob/4.x/doc/pattern.png exising one?
- what about adding also a `pdf` version?
- adding an image of the different calibration boards for illustration in this tutorial (https://docs.opencv.org/4.5.5/da/d0d/tutorial_camera_calibration_pattern.html)?
- this tutorial (https://docs.opencv.org/4.5.5/dc/d43/tutorial_camera_calibration_square_chess.html) is obsolete and should be removed or prepended with an `obsolete` keyword in the title and at the beginning of the page in my opinion
- keep the current tools, but add more complete (video?) tutorials, with good advices about how to select good calibration views?

## Leveraging some existing calibration tools?

- **no need to reinvent the wheel**
- what about a collaboration to refresh this tool (https://april.eecs.umich.edu/papers/details.php?name=richardson2013iros / https://april.eecs.umich.edu/wiki/Camera_suite)?
  - refresh of the tool if needed, or complete rewrite of the tool in a modern language if needed?
  - add corresponding tutorials and guidelines in the OpenCV documentation site and let the hosting of the tool on a third party site?
- other tools are:
  -  https://github.com/paroj/pose_calib
  - https://github.com/oliver-batchelor/multical (from Github suggestion 😃 )
  - on Github, [...]
  - depending on the license term / code language, can be a collaboration between both parties?
- where to host the calibration app binary?, an easy procedure to download and launch the tool is important

## Updating the current [interactive camera calibration app](https://github.com/opencv/opencv/tree/master/apps/interactive-calibration)?

- **warning, no experience with it**
- add features such as:
  - possibility to manually remove bad images (blurry, redundant, ...)?
  - possibility to manually or automatically remove images with large projection error? (is it a good idea? will not introduce a bias rather?)
  - does it handle stereo calibration, fisheye calibration?
  - add a GUI similar to the Matlab tool (https://www.mathworks.com/help/vision/ug/using-the-single-camera-calibrator-app.html / https://www.mathworks.com/help/vision/ug/using-the-stereo-camera-calibrator-app.html)?

![image](https://user-images.githubusercontent.com/8035162/161399641-c544ef70-7a2a-40ca-a159-1108fc295df6.png)

![image](https://user-images.githubusercontent.com/8035162/161399644-ca9152ca-96ff-480e-b6a4-859e31670279.png)

## Misc

- what about refreshing the tools, tutorials to use modern calibration patterns such as the `AprilTag` or `ChArUco` boards?
  - multiple advantages such as better detection rate?, better precision of the corners location extraction?, no need to view the whole board --> good for stereo calibration and the image corners
- what about this new `findChessboardCornersSB()` (https://github.com/opencv/opencv/pull/16625)?
  - update the tutorials to use this calibration board?
  - add corresponding tutorials?
  - add the corresponding `png` / `pdf` calibration board images / links in the tutorials?

---

Hope this helps 🥂

**Comment by s-trinh** [comment]
An interesting camera calibration tool written by @dkogan is https://github.com/dkogan/mrcal

See [A tour of mrcal](http://mrcal.secretsauce.net/tour.html).

It contains interesting analysis tools such as:

![image](https://user-images.githubusercontent.com/8035162/170143946-10fe3321-ca0c-4566-88f1-a56ace664cc8.png)

**Comment by dkogan** [comment]
Yeah, you should be using my tools (mrcal). The other ones all repackage the same flawed ideas, resulting in the same problems. mrcal resolves these conclusively, with new research. Ask me for help, if you need it.

**Comment by scottlawsonbc** [comment]
I had a great experience using @dkogan's mrcal tool. The visualizations are really nice and helped me improve my workflow for calibrating cameras. In particular, the visualizations showing me the distribution of detected points prompted me to retake calibration images and improve my calibration procedure overall. There are a few gems that would be great to include in opencv. I appreciate the more scientific and disciplined approach to modeling camera calibration that you can find in mrcal.
